blockmanager.go

```diff
@@ -3044,3 +3091,167 @@ func (l *lightHeaderCtx) RelativeAncestorCtx(
+	if err != nil {
+		log.Errorf("error while setting height for sideload--- skipping sideloading: "+
+			"%v", err)
+
+		return
+	}
+
+	// Start batch writer to write headers into DB at specified batch size and ticker duration.
+	b.blockHeaderBatchWriter.Start()
+	defer b.blockHeaderBatchWriter.Stop()
+	for {
+		select {
+		case <-b.quit:
+			return
+		default:
+			// Request header
+			header, headerErr := b.sideLoadBlkHdrReader.NextHeader()
+			// If any error occurs while fetching headers that does not indicate
+			// an end of file, return immediately.
+			if headerErr != nil {
+				if headerErr == chaindataloader.ErrEndOfFile {
+					log.Infof("Successfully completed sideloading")
+					return
+				}
+
+				log.Errorf("error while fetching headers -- skipping sideloading %v", err)
+				return
+			}
+
+			var (
+				node     *headerlist.Node
+				prevNode *headerlist.Node
+			)
+			// Ensure there is a previous header to compare against.
+			prevNodeEl := b.headerList.Back()
+			if prevNodeEl == nil {
+				log.Warnf("side load - Header list does not contain a previous" +
+					"element as expected -- exiting side load")
+
+				return
+			}
+
+			node = &headerlist.Node{Header: *header}
+			prevNode = prevNodeEl
+			node.Height = prevNode.Height + 1
+
+			if !b.cfg.sideLoad.SkipVerify {
+				valid, err := b.verifyBlockHeader(header, *prevNode)
+				if err != nil || !valid {
+					log.Debugf("Side Load- Did not pass verification at height %v"+
+						"-- rolling back to last verified checkpoint and skipping sideload", node.Height)
+
+					prevCheckpoint := b.findPreviousHeaderCheckpoint(
+						node.Height,
+					)
+
+					log.Infof("Rolling back to previous validated "+
+						"checkpoint at height %d/hash %s",
+						prevCheckpoint.Height,
+						prevCheckpoint.Hash)
+
+					// It is possible that the number of headers in the batch writer's buffer is less than the
+					// BlkHeaderWriteBatch and so these headers have not been written into the DB at this time.
+					// We wait for the DBTickerBlkHeaderWriteDuration to expire and write these headers into the
+					// DB before we roll back which should take approximately twice the DBTickerBlkHeaderWriteDuration.
+					// This ensures that these headers are wiped off, during the rollback, in the batch writer's buffer
+					// and in the DB as well.
+					time.Sleep(DBTickerBlkHeaderWriteDuration * 2)
+					err = b.rollBackToHeight(uint32(prevCheckpoint.Height))
+					if err != nil {
+						panic(fmt.Sprintf("Rollback failed: %s", err))
+						// Should we panic here?
+					}
+					tipHeader, height, err := b.cfg.BlockHeaders.ChainTip()
+					if err != nil {
+						return
+					}
+
+					b.headerList.ResetHeaderState(headerlist.Node{
+						Height: int32(height),
+						Header: *tipHeader,
+					})
+					return
+				}
+
+				// Verify checkpoint only if verification is enabled.
+				if b.nextCheckpoint != nil &&
+					node.Height == b.nextCheckpoint.Height {
```

Comment:
We should feed all this through our normal header processing logic. Some refactoring might be needed.

Reply:
You mean refactoring the `handleheaders` function?

Reply:
For regular block headers, I think the sideloading should be in `blockManager.Start()` after the `blockHandler` goroutine has started. Then it can call `handleHeadersMsg` with the set of block headers.  The `headersMsg` struct might need to be modified since there's no `peer`. This also gets rid of the code duplication

For filter headers, something similar can be done where the sideloading is in `Start()`.

Reply:
IMO it's better to have this be distinct, as then we don't need to increase the set of responsibilities of the `blockManager`. 

Logic similar to the following can validate all teh headers at oce: 
```go
	for height := uint32(1); height <= bestBlockHeight; height++ {
		if currHeader.PrevBlock != prevHeader.BlockHash() {
			return fmt.Errorf("block header at height %d does "+
				"not refrence previous hash %v", height,
				prevHeader.BlockHash().String())
		}

		parentHeaderCtx := newLightHeaderCtx(
			int32(height-1), &prevHeader, headerStore, nil,
		)

		skipDifficulty := blockchain.BFFastAdd
		err = blockchain.CheckBlockHeaderContext(
			&currHeader, parentHeaderCtx, skipDifficulty, chainCtx,
			true,
		)
		if err != nil {
			return fmt.Errorf("error checking block %d header "+
				"context: %w", height, err)
		}

		err = blockchain.CheckBlockHeaderSanity(
			&currHeader, chainCtx.params.PowLimit, timeSource,
			skipDifficulty,
		)
		if err != nil {
			return fmt.Errorf("error checking block %d header "+
				"sanity: %w", height, err)
		}

		// TODO: Validate checkpoint and filter headers.
		prevHeader = currHeader

		if height > 0 && height%100_000 == 0 {
			log.Debugf("Validated %d headers", height)
		}
	}
```

We'd then: 
  1. Read a chunk of `N` headers (say 2k or so) from the side loader source. 
  2. Validate them all at once. 
  3. Write them all to disk directly.

Reply:
Cool, I though we did not want to duplicate the functionality of the `handleheaders` function https://github.com/lightninglabs/neutrino/pull/285#discussion_r1377989999